### ISSUE TITLE
memory fault on list-bookmarks

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2062]
If I run `list-bookmarks` over my 3000+ bookmark file, I get

```log
Unhandled memory fault at #x7F429F5DC010.
   [Condition of type SB-SYS:MEMORY-FAULT-ERROR]

Restarts:
 0: [ABORT] abort thread (#<THREAD "Nyxt renderer thread" RUNNING {1010D28AA3}>)

Backtrace:
 0: (SB-SYS:MEMORY-FAULT-ERROR #<unused argument> #.(SB-SYS:INT-SAP #X7F429F5DC010))
 1: ("foreign function: call_into_lisp")
 2: ("foreign function: funcall2")
 3: ("foreign function: handle_memory_fault_emulation_trap")
 4: ("foreign function: #x4176CE")
 5: ((LAMBDA (&REST GIR::ARGS-IN) :IN GIR::BUILD-FUNCTION))
 6: ((FLET NYXT::MAIN-FUNC :IN NYXT:FFI-INITIALIZE))
 7: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
 8: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
 9: ((FLET "WITHOUT-INTERRUPTS-BODY-10" :IN SB-THREAD::RUN))
10: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
11: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-THREAD::RUN))
12: (SB-THREAD::RUN)
13: ("foreign function: call_into_lisp")
14: ("foreign function: funcall1")
```

It works on a small 2-entry bookmark file.
The 3000-entry bookmark file works on 2.2.4.

@aartaka This may be due to some change in the internal-page system.  Any idea what to investigate?

**Comment by aartaka** [comment]
The first guess is that we cannot allocate a C string long enough to fit the HTML for 3000+ bookmarks. I guess the only way we can work around this is by adding pagination/lazy loading to `list-bookmarks`.

**Comment by Ambrevar** [comment]
Where does this allocation happen?
Why is this a problem now and not before?

**Comment by aartaka** [comment]
Allocation happens in `cl-webkit`, in `uri-scheme-processed` callback.

But then the backtrace would at least mention `cl-webkit`. Maybe it's something else. Does it fail if you do `set-url-from-bookmark` or other bookmark-related commands?

**Comment by Ambrevar** [comment]
No, the rest of the bookmark commands work.

**Comment by Ambrevar** [comment]
Proposed fix: don't populate the buffer via JS, instead just call a Lisp function that will be in charge of populating the buffer.

**Comment by Ambrevar** [comment]
Maybe use https://github.com/wlbr/cl-marshal ?

**Comment by aartaka** [comment]
Can reproduce with the bookmarks file containing 3000 duplicate entries like
```
(:url "https://webkitgtk.org/reference/webkit2gtk/stable/index.html" :title "WebKitGTK Reference Manual: WebKitGTK Reference Manual" :date "2021-11-22T05:13:25.505081Z")
```

**Comment by aartaka** [comment]
UPD: Can reproduce with 930+ entries like the one above.

**Comment by aartaka** [comment]
... and 40000+ bookmarks is being deserialized just fine. 

So, I guess, we have two problems:
- Serialization of big quantities of data into HTML (`spinneret` and WebKitGTK).
- Deserialization of deeply nested data from e.g. history (`cl-prevalence`).

The first problem is solvable via optimizations on the side of `spinneret` or swithing to a more space-preserving library.

The second is either 
- optimizing `cl-prevalence`, 
- switching to another (de)serialization library (`cl-marshal`?),
- or simplifying the `htree` structure to reduce nesting.

**Comment by Ambrevar** [comment]
Wait, or you saying that it fails with 930+ bookmarks when serializing, but it can deserialize 40000+ bookmarks?

In the original post, it's failing in `list-bookmars`, thus in deserializing 3000 bookmarks.  Thus nesting is not to blame here, there is another issue which I presume is with the data transfer between WebKitGTK and Lisp.

**Comment by aartaka** [comment]
This should've been confusing, sorry. 
- 930+ bookmarks exhaust the memory (I suppose) when being serialized by Spinneret.
- 40000+ bookmarks are being deserialized into lispy structures just fine by CL-Prevalence.

**Comment by aartaka** [comment]
> In the original post, it's failing in `list-bookmars`, thus in deserializing 3000 bookmarks. Thus nesting is not to blame here, there is another issue which I presume is with the data transfer between WebKitGTK and Lisp.

Does it also fail in `set-url-bookmark` (or whatever the name is, not at my machine atm)?

**Comment by aartaka** [comment]
It doesn't. So bookmarks are being deserialized fine from the file, and only fail to work when translated to HTML.

**Comment by aartaka** [comment]
b7dd2f1e0888b4bac2d3af99889d77f0896d6634 should defer the dreadful appearance of segfaults. However, it's not _the_ solution to the fact that C-side (it seems) doesn't take more that 1000000 characters at a time. I've tested, and Spinneret/Lisp is fine with allocating this much (i.e. the body of `list-bookmarks` terminates normally), while WebKit/CFFI seems to choke on it (causing a memory fault).

While b7dd2f1e0888b4bac2d3af99889d77f0896d6634 is reducing the size of `list-bookmarks` HTML by more than three times (see commit message for details), it would still fail at some 3500+ bookmarks, which is still too little to bear with. I am on the side of pagination as _the_ fix for `list-bookmarks` being overly greedy.

**Comment by Ambrevar** [comment]
But we could do easier, as I suggested earlier: have the C side just
handle what to do, and have the Lisp side handle everything else, namely
to populate the HTML buffer either directly or via some Parenscript.

**Comment by aartaka** [comment]
Naaaaaaah...

I don't like our stateful Parenscript-based page modification that was there before the internal pages change. It's not exactly cacheable, and is dependent on JavaScript being enabled. I have a feeling that it's better to go with the renderer-friendly static internal pages and introduce pagination/hierarchy to handle huge pieces of data, especially now that internal pages have history.

**Comment by Ambrevar** [comment]
But if we implement my suggestion just for list-bookmarks, it does
no harm to the rest of the browser (and we can still implement
pagination later).

No need for JS either: we can populate the HTML directly (using Spinneret).

**Comment by aartaka** [comment]
Then I must've misunderstood you: how do you suggest populating the page with Spinneret?

**Comment by Ambrevar** [comment]
You can load HTML directly for instance with

https://webkitgtk.org/reference/webkit2gtk/stable/WebKitWebView.html#webkit-web-view-load-html

**Comment by aartaka** [comment]
Yes, that's indeed an option. However, `webkit-web-view-load-html` has some drawbacks:
- It's agressively cached, so page reload returns you exactly the same content it set, unless you override it or reload without cache.
- It requires a URL that invokes all the callbacks and is saved to history.
  - Providing no URL is not an option -- `about:blank` is used then.
- It still is a solution relying on the C-side taking whatever we give. While I haven't tested it, it could segfault on huge inputs in the same fashion as the internal schemes apparatus does.

**Comment by jmercouris** [comment]
Pagination is not necessary. Notice that I implemented the function `ffi-buffer-get-document`:

```
(define-ffi-generic ffi-buffer-get-document (buffer)
  (:method ((buffer buffer))
    (pflet ((get-html (start end)
                      (ps:chain document document-element |innerHTML| (slice (ps:lisp start)
                                                                             (ps:lisp end))))
            (get-html-length ()
                             (ps:chain document document-element |innerHTML| length)))
      (with-current-buffer buffer
        (let ((slice-size 10000))
          (reduce #'str:concat
                  (loop for i from 0 to (truncate (get-html-length)) by slice-size
                        collect (get-html i (+ i slice-size)))))))))
```

the key is to define some sort of slice size and keep adding things via JS. This way we can circumvent the limitations of buffer sizes in C.

**Comment by Ambrevar** [comment]
Hmm, isn't it precisely what Artyom meant with "pagination"?  Maybe the
more popular term here would be "buffering".

**Comment by aartaka** [comment]
No, injecting all the data piece-by-piece (buffering, John's suggestion), or breaking data into separate pages (pagination, my suggestion) are slightly different, although the general idea is similar -- breaking the data into pieces to ease renderer's life :)

**Comment by Ambrevar** [comment]
Can you elaborate?

**Comment by jmercouris** [comment]
The difference is that my suggestion would all of the data to the page at once, Artyom's suggestion would require user action to go to the "next page". Imagine if all Google results appeared on one very long page vs clicking to get to the next page. This is the difference.

**Comment by Ambrevar** [comment]
@aartaka fixed the bug in cl-webkit 3.5.2.
